pkg/virt-controller/watch/replicaset_test.go

```diff
@@ -0,0 +1,506 @@
+		var recorder *record.FakeRecorder
+
+		sync := func(stop chan struct{}) {
+			go vmInformer.Run(stop)
+			go rsInformer.Run(stop)
+			Expect(cache.WaitForCacheSync(stop, vmInformer.HasSynced, rsInformer.HasSynced)).To(BeTrue())
+		}
+
+		BeforeEach(func() {
+			stop = make(chan struct{})
+			ctrl = gomock.NewController(GinkgoT())
+			virtClient = kubecli.NewMockKubevirtClient(ctrl)
+			vmInterface = kubecli.NewMockVMInterface(ctrl)
+			rsInterface = kubecli.NewMockReplicaSetInterface(ctrl)
+
+			vmSource = framework.NewFakeControllerSource()
+			rsSource = framework.NewFakeControllerSource()
+			vmInformer = cache.NewSharedIndexInformer(vmSource, &v1.VirtualMachine{}, 0, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+			rsInformer = cache.NewSharedIndexInformer(rsSource, &v1.VirtualMachineReplicaSet{}, 0, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+			recorder = record.NewFakeRecorder(100)
+
+			controller = NewVMReplicaSet(vmInformer, rsInformer, recorder, virtClient, uint(10))
+		})
+
+		It("should create missing VMs and increase the replica count", func() {
+			vm := v1.NewMinimalVM("testvm")
+			vm.ObjectMeta.Labels = map[string]string{"test": "test"}
+			rs := ReplicaSetFromVM("rs", vm, 3)
+
+			expectedRS := clone(rs)
+			expectedRS.Status.Replicas = 0
+			expectedRS.Status.ReadyReplicas = 0
+
+			rsSource.Add(rs)
+
+			sync(stop)
+
+			virtClient.EXPECT().VM(vm.ObjectMeta.Namespace).Return(vmInterface).AnyTimes()
+			vmInterface.EXPECT().Create(gomock.Any()).Times(3).Do(func(arg interface{}) {
+				Expect(arg.(*v1.VirtualMachine).ObjectMeta.GenerateName).To(Equal("testvm"))
+			}).Return(vm, nil)
+			virtClient.EXPECT().ReplicaSet(vm.ObjectMeta.Namespace).Return(rsInterface)
+			rsInterface.EXPECT().Update(expectedRS)
+
+			controller.Execute()
+
+			expectEvent(recorder, SuccessfulCreateVirtualMachineReason)
+			expectEvent(recorder, SuccessfulCreateVirtualMachineReason)
+			expectEvent(recorder, SuccessfulCreateVirtualMachineReason)
+		})
+
+		It("should create missing VMs in batches of a maximum of 10 VMs at once", func() {
+			vm := v1.NewMinimalVM("testvm")
+			vm.ObjectMeta.Labels = map[string]string{"test": "test"}
+			rs := ReplicaSetFromVM("rs", vm, 15)
+
+			rsSource.Add(rs)
+
+			sync(stop)
+
+			virtClient.EXPECT().VM(vm.ObjectMeta.Namespace).Return(vmInterface).AnyTimes()
+			virtClient.EXPECT().ReplicaSet(vm.ObjectMeta.Namespace).Return(rsInterface).AnyTimes()
+			rsInterface.EXPECT().Update(gomock.Any()).AnyTimes()
+
+			invocations := 0
+			// Count up on every invocation, so that we can measure how often this was called on one invocation
+			vmInterface.EXPECT().Create(gomock.Any()).Times(15).Do(func(arg interface{}) {
+				Expect(arg.(*v1.VirtualMachine).ObjectMeta.GenerateName).To(Equal("testvm"))
+			}).Return(vm, nil).Do(func(obj interface{}) {
+				invocations++
+			})
+
+			// Check if only 10 are created
+			controller.Execute()
+			Expect(invocations).To(Equal(10))
```

Comment:
Are the mock callbacks somehow execute asynchronously?  This regression test failed during one of its runs, but passes at other times.

Reply:
Seems so. I am now just specifying the expected amount of invocations. That works.